Allow to init a crystal app/lib in an empty directory

bew opened the pull request:
I usually create a github repository, then clone it and finally want to use `crystal init lib my_lib .` to initialize a lib project in the current directory. 
This previously failed with error `file or directory . already exists`.

This PR simply allow the directory to be `.` (the current directory).

RX14 commented:
I think we should check that the directory is empty (apart from .git) beforehand.

bew commented:
> I think we should check that the directory is empty (apart from .git) beforehand.

@RX14 instead of checking that the directory exists or not you mean ?

RX14 commented:
@bew yes.

Fryguy commented:
> I think we should check that the directory is empty (apart from .git) beforehand.

If done generically this can also handle the case where one creates a directory beforehand.

```bash
> mkdir my_new_app
> crystal init app my_new_app
file or directory my_new_app already exists  # Grrrrrrrr
> rm -rf my_new_app
> crystal init app my_new_app
```

bew commented:
Updated, allowing to `crystal init` in an empty directory (only the .git/ directory is allowed).

RX14 commented:
@bew What do you mean by "handle other VCS"? `git init` in an existing repository is harmless.

bew commented:
I mean, I use this functionnality this way:
```
$ git clone url/to/my_repo
$ cd my_repo
$ crystal init lib my_lib .
```
This will make a `.git` directory that the current PR ignores, but what if the user uses mercury, the directory will contain a `.hg` directory, and the current PR will say that the directory is not empty. Should we ignore `.hg/` directory too (and possibly other VCS (Version Control System) directories?)

straight-shoota commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -64,12 +65,29 @@ module Crystal
 
+    def self.directory_valid?(directory)
+      if File.file?(directory)
+        puts "file #{directory} already exists"
```
The error message should not only contain the failure cause, but also what this means in context. Like: `cannot initialize Crystal project: #{directory} it is a file`

straight-shoota commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -64,12 +65,29 @@ module Crystal
+      if Dir.exists?(directory)
+        Dir.foreach(directory) do |f|
+          unless ALLOWED_DIRECTORY_FILES.includes?(f)
+            puts "directory #{directory} is not empty (only .git/ is allowed)"
```
Similarly: `cannot initialize Crystal project: #{directory} is not empty (only .git/ folder is allowed)`

straight-shoota commented:
Hm, maybe we could ignore all files and folders starting with a `.`? Or why fail at all if there are already contents in the directory?

bew commented:
Thanks @straight-shoota, I've updated with better error messages.

> why fail at all if there are already contents in the directory?

I think it is to prevent overwriting an existing project files.
Maybe we could make a check, verifying existing files, and stop the command when it will overwrite some files. If the user wants to reset the files, we could add a `--force` flag?

straight-shoota commented:
Or by default the init command could ask what to do if a spacific path already exists.

RX14 commented:
We can't just use the "all paths starting with ." rule as `crystal init` creates dotfiles of it's own. I like the idea of erroring out only if it would overwrite, however this would have to be atomic: erroring out halfway is annoying. I think having all this code to atomically init a crystal project would be a bit complex, I think the current heuristic of just allowing `.git` is useful for covering the common usecase established in this PR.

straight-shoota commented:
@RX14 I don't think this should be too complex... the generated paths are all known as `#full_path` on each view. So this should be sufficient:
```crystal
# add this to InitProject#run
raise if views.any? { |view| File.exists?(view.full_path) }
```
Or am I missing something?

RX14 commented:
@straight-shoota I admit, I hadn't actually looked at the `crystal init` code before saying it was complex. Your proposed solution looks fine.

straight-shoota commented:
@bew what about the idea to only fail if a generated path already exists? This would make it play nice with any source control system, other project initialization tools etc. or even you have already begun to add code before initializing.

Fryguy commented:
https://github.com/crystal-lang/crystal/pull/4808 is merged, so now I think you can use Dir.children to eliminate the "." and ".." bits without having to specify here.

bew commented:
@straight-shoota I like the idea, I'm trying to implement it right now!
@Fryguy thanks for the mention! but as soon as I finish the implementation of @straight-shoota's idea, the `Dir#foreach` I currently use won't be necessary anymore.

bew commented:
Given:
```
$ mkdir dir && cd dir
$ echo "# my_readme" > README.md
```
Using `crystal init lib my_project .` will fail saying that `README.md` would be overwritten.

What if we want to:
- overwrite the file (using a `--force` or `--overwrite-existing` flag?)
- skip the file but generate the other files (using a `--skip-existing` flag?)

What do you think?

straight-shoota commented:
As this command will usually be run manually by a developer on a terminal I'd suggest to handle this interactively by default and ask the user what to do: overwrite, skip, abort. Of course there should also be flags for that.

While this would generally be nice, it doesn't need to be too complicated. I guess it would also be sufficent to abort by default and allow `--force` and `--skip-existing` flags to define the behaviour in case of collision.

bew commented:
Done! I've added the flags `--force` & `--skip-existing`.

I've also refactored the specs with 2 helpers. It does a lot of diffs caused by an indentation change. To keep this PR's diff as clean as possible I can make another PR with those refactors, to be merged before this one. WDYT?

straight-shoota commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -31,13 +30,26 @@ module Crystal
           exit
         end
 
+        opts.on("--force", "force overwrite existing files") do
```
It would be nice to have `-f` as as shorthand. Don't know about the other flag...

bew replied:
What about `-s` for `--skip-existing` ?

straight-shoota commented:
@bew could you remove the commits of #4915 from this or rebase on that branch?

bew commented:
@straight-shoota I think #4915 will soon be merged, I'll rebase on it when it happen!

bew commented:
#4915 has just been merged!
I rebased & squashed the commits here, it should be a lot easier to see the real changes thanks to #4915.

bew commented:
I think we should wait for #5520 merge before merging this PR, I think it would be cleaner in this order.

RX14 commented:
This needs an update now #5520 is merged.

bew commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -155,18 +180,36 @@ module Crystal
+              STDERR.puts "   #{"file".colorize(:red)} #{view.full_path} #{"already exist".colorize(:red)}"
+            end
+            STDERR.puts "You can use --force to overwrite those files,"
+            STDERR.puts "or --skip-existing to skip existing files and generate the others."
```
How can I handle these messages properly using an exception?
I was thinking about using a custom exception with a custom field of the list of conflicting file paths, then in `compiler/crystal/command.cr` rescue this exception and print this long message using the custom field.

WDYT?

bew replied:
So I implemented this, it looks ok, and works perfectly ;)

bew commented:
Rebased and refactored a bit to have simpler logic in `overwrite_checks`

asterite commented:
Looks good, though I would leave the exception handling logic inside tools init. Otherwise is looks a bit disconnected.

bew commented:
@asterite what do you mean?

asterite commented on src/compiler/crystal/command.cr:
```diff
@@ -159,8 +159,16 @@ class Crystal::Command
   private def init
     begin
```
I mean this whole `begin`/`rescue` should be in `src/compiler/crystal/tools/init.cr`, probably inside the `Init.run` method.

bew replied:
Thanks for the precision, should be good now!

bew commented:
For the conflict resolution, in addition to `--force` & `--skip-existing` I was thinking about an `--interactive` mode too, where the user could be prompted `overwrite "some_file.cr" ? [y/N]` for each conflicting files.

But I can probably do that in another PR?
